plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -221,21 +219,70 @@ func findNodesThatFit(
 }
 
 // Checks whether node with a given name and NodeInfo satisfies all predicateFuncs.
-func podFitsOnNode(pod *v1.Pod, meta interface{}, info *schedulercache.NodeInfo, predicateFuncs map[string]algorithm.FitPredicate) (bool, []algorithm.PredicateFailureReason, error) {
-	var failedPredicates []algorithm.PredicateFailureReason
-	for _, predicate := range predicateFuncs {
-		fit, reasons, err := predicate(pod, meta, info)
-		if err != nil {
-			err := fmt.Errorf("SchedulerPredicates failed due to %v, which is unexpected.", err)
-			return false, []algorithm.PredicateFailureReason{}, err
+func podFitsOnNode(pod *v1.Pod, meta interface{}, info *schedulercache.NodeInfo, predicateFuncs map[string]algorithm.FitPredicate,
+	ecache *EquivalenceCache) (bool, []algorithm.PredicateFailureReason, error) {
+	var (
+		equivalenceHash  uint64
+		failedPredicates []algorithm.PredicateFailureReason
+		eCacheAvailable  bool
+		invalid          bool
+		fit              bool
+		reasons          []algorithm.PredicateFailureReason
+		err              error
+	)
+	if ecache != nil {
+		// getHashEquivalencePod will return immediately if no equivalence pod found
+		equivalenceHash = ecache.getHashEquivalencePod(pod)
+		eCacheAvailable = (equivalenceHash != 0)
+	}
+	for predicateKey, predicate := range predicateFuncs {
+		// If equivalenceCache is available
+		if eCacheAvailable {
+			// PredicateWithECache will returns it's cached predicate results
+			fit, reasons, invalid = ecache.PredicateWithECache(pod, info.Node().GetName(), predicateKey, equivalenceHash)
 		}
-		if !fit {
+
+		if !eCacheAvailable || invalid {
+			// we need to execute predicate functions as normal since equivalence cache does not work
+			if failedPredicates, err = processPredicateNormally(predicateKey, predicate, pod, meta, info, failedPredicates, eCacheAvailable, ecache, equivalenceHash); err != nil {
+				return false, []algorithm.PredicateFailureReason{}, fmt.Errorf("SchedulerPredicates failed due to %v, which is unexpected.", err)
+			}
+		} else if !fit {
+			// eCache is available and valid, and predicates result is unfit, record the fail reasons
```

Comment:
Yeah, that's my original version, while it have been changed for multiple times to current version ... refactoring